geom/twkb_parser.go

```diff
@@ -0,0 +1,562 @@
+			// So determine if the first and final coords differ.
+			finalPointDiffersFromFirst := false
+			for d := 0; d < p.dimensions; d++ {
+				first := coords[0+d]
+				final := coords[(numPoints-1)*p.dimensions+d]
+				if first != final {
+					finalPointDiffersFromFirst = true
+					break
+				}
+			}
+
+			if finalPointDiffersFromFirst {
+				// Append first point again, to close the ring.
+				for d := 0; d < p.dimensions; d++ {
+					coords = append(coords, coords[0+d])
+				}
+			}
+		}
+		ls, err := NewLineString(NewSequence(coords, p.ctype), p.opts...)
+		if err != nil {
+			return Polygon{}, p.annotateError(err)
+		}
+		rings = append(rings, ls)
+	}
+	poly, err := NewPolygon(rings, p.opts...)
+	return poly, p.annotateError(err)
+}
+
+func (p *TWKBParser) parseMultiPoint() (MultiPoint, error) {
+	if p.isEmpty {
+		return NewMultiPoint(nil, p.opts...), nil
+	}
+	return p.nextMultiPoint()
+}
+
+func (p *TWKBParser) nextMultiPoint() (MultiPoint, error) {
+	numPoints, err := p.parseUnsignedVarint()
+	if err != nil {
+		return MultiPoint{}, p.parserError("cannot unmarshal num points varint malformed")
+	}
+	_, err = p.parseIDList(int(numPoints))
+	if err != nil {
+		return MultiPoint{}, err
+	}
+	var pts []Point
+	for i := 0; i < int(numPoints); i++ {
+		pt, err := p.nextPoint()
+		if err != nil {
+			return MultiPoint{}, err
+		}
+		pts = append(pts, pt)
+	}
+	return NewMultiPoint(pts, p.opts...), nil
+}
+
+func (p *TWKBParser) parseMultiLineString() (MultiLineString, error) {
+	if p.isEmpty {
+		return NewMultiLineString(nil, p.opts...), nil
+	}
+	return p.nextMultiLineString()
+}
+
+func (p *TWKBParser) nextMultiLineString() (MultiLineString, error) {
+	numLineStrings, err := p.parseUnsignedVarint()
+	if err != nil {
+		return MultiLineString{}, p.parserError("cannot unmarshal num linestrings varint malformed")
+	}
+	_, err = p.parseIDList(int(numLineStrings))
+	if err != nil {
+		return MultiLineString{}, err
+	}
+	var lines []LineString
+	for i := 0; i < int(numLineStrings); i++ {
+		ls, err := p.nextLineString()
+		if err != nil {
+			return MultiLineString{}, err
+		}
+		lines = append(lines, ls)
+	}
+	return NewMultiLineString(lines, p.opts...), nil
+}
+
+func (p *TWKBParser) parseMultiPolygon() (MultiPolygon, error) {
+	if p.isEmpty {
+		return NewMultiPolygon(nil, p.opts...)
+	}
+	return p.nextMultiPolygon()
+}
+
+func (p *TWKBParser) nextMultiPolygon() (MultiPolygon, error) {
+	numPolygons, err := p.parseUnsignedVarint()
+	if err != nil {
+		return MultiPolygon{}, p.parserError("cannot unmarshal num polygons varint malformed")
+	}
+	_, err = p.parseIDList(int(numPolygons))
+	if err != nil {
+		return MultiPolygon{}, err
+	}
+	var polys []Polygon
+	for i := 0; i < int(numPolygons); i++ {
+		poly, err := p.nextPolygon()
+		if err != nil {
+			return MultiPolygon{}, err
+		}
+		polys = append(polys, poly)
+	}
+	mp, err := NewMultiPolygon(polys, p.opts...)
+	return mp, p.annotateError(err)
```

Comment:
I noticed that `annotateError` is used in some places but not others. 

Would it make sense to annotate the error once in `UnmarshalTWKB` (bottom of the call stack) and instead of `p.parserError`, return the built-in error via `fmt.Errorf(...)` or `errors.New(...)` instead? It would help reduce potential duplication of the byte position annotation in error messages.

Reply:
I was trying to use it only in return paths where the byte position wasn't otherwise being reported. Good point about annotating in one spot, I'll try to tidy that up.